news/utils.py

```diff
@@ -0,0 +1,138 @@
+import re
+import requests
+from bs4 import BeautifulSoup, NavigableString
+from usp.tree import sitemap_tree_for_homepage
+
+def build_url_index(base_url, keyword=None):
+
+    if keyword is None:
+        keyword = [""]
+    elif isinstance(keyword, str):
+        keyword = [keyword]
+
+    tree = sitemap_tree_for_homepage(base_url)
+    page_index = [page.url for key in keyword for idx, page in enumerate(tree.all_pages()) if key in page.url]
+    return page_index
+
+
+def get_text_from_page(url=None, html_path=None, tag="article", attrs=None):
+
+    assert bool(url is not None) != bool(html_path is not None)
+
+    if url is not None:
+        # headers = {'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/39.0.2171.95 Safari/537.36'}
+        headers = {'User-Agent': 'My User Agent 1.0'}
+        page = requests.get(url, headers=headers)
+        soup = BeautifulSoup(page.content, "html.parser")
+    elif html_path is not None:
+        with open(html_path, "rb") as fp:
+            soup = BeautifulSoup(fp, "html.parser")
+
+    text = [soup.title.get_text() if soup.title else ""]
+
+    # Search for author
+    if byline := soup.find(class_=re.compile("byline")):
+        text.append(byline.get_text().strip())
+    elif byline := soup.find(class_=re.compile("post-author")):
```

Comment:
I confirmed it myself, but can you add a comment like

```
# Calling `re.compile(...)` repeatedly in a this processing function (once for each document) is ok because the resulting compiled re object is cached and reused
```

Otherwise I'm going to forget and wonder if repeated calls is an issue in the future lol.